File: antigravity-backend/services/context_engine.py
from services.vector_memory import search_memory


class ContextEngine:
    def __init__(self):
        self.min_relevance_threshold = 0.7

    async def get_context(self, task: str, n_results: int = 5):
        related = search_memory(task, n_results)
        return self.format_context(task, related)

    def format_context(self, task: str, data: dict):
        related_memories = []
        relevance_scores = []

        for i, doc in enumerate(data.get("documents", [])):
            distance = data.get("distances", [1])[i]
            relevance = 1 - distance

            if relevance >= self.min_relevance_threshold:
                related_memories.append(doc)
                relevance_scores.append(relevance)

        context = "\n".join([f"- {item}" for item in related_memories])
        if not context:
            context = "* Nenhuma memória relevante encontrada."

        return {
            "task": task,
            "related_memories": related_memories,
            "context": context,
            "relevance_scores": relevance_scores
        }

    def set_threshold(self, threshold: float):
        self.min_relevance_threshold = threshold


async def get_context(task: str, n_results: int = 5):
    engine = ContextEngine()
    return await engine.get_context(task, n_results)

File: services/context_engine.ts
import { searchMemory, type MemorySearchResult } from "./vector_memory";

export interface ContextResult {
  task: string;
  relatedMemories: string[];
  context: string;
  relevanceScores: number[];
}

export class ContextEngine {
  private minRelevanceThreshold = 0.7;

  async getContext(task: string, nResults: number = 5): Promise<ContextResult> {
    const searchResult = await searchMemory(task, nResults);
    
    return this.formatContext(task, searchResult);
  }

  private formatContext(
    task: string,
    data: { ids: string[]; documents: string[]; distances: number[] }
  ): ContextResult {
    const relatedMemories: string[] = [];
    const relevanceScores: number[] = [];

    data.ids.forEach((id, index) => {
      const distance = data.distances[index] ?? 1;
      const relevance = 1 - distance;
      
      if (relevance >= this.minRelevanceThreshold) {
        relatedMemories.push(data.documents[index]);
        relevanceScores.push(relevance);
      }
    });

    let context = "";
    relatedMemories.forEach((item) => {
      context += `- ${item}\n`;
    });

    if (!context) {
      context = "* Nenhuma memória relevante encontrada.\n";
    }

    return {
      task,
      relatedMemories,
      context,
      relevanceScores,
    };
  }

  setThreshold(threshold: number): void {
    this.minRelevanceThreshold = threshold;
  }

  async getFullContext(task: string): Promise<string> {
    const context = await this.getContext(task);
    return context.context;
  }
}

export const contextEngine = new ContextEngine();

export async function getContext(task: string): Promise<ContextResult> {
  return contextEngine.getContext(task);
}

export async function getContextString(task: string): Promise<string> {
  return contextEngine.getFullContext(task);
}